assets/js/util/report-validation.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Reporting API validation utilities.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Validates data that can be either string or object of the certain type, or array of them.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} data The data to check.
+ * @param {Function} objectVerificationCallback The callback to verify an object.
+ * @return {boolean} TRUE if data is valid, otherwise FALSE.
+ */
+function isValidStringsOrObjects( data, objectVerificationCallback ) {
+	if ( typeof data === 'string' ) {
+		return true;
+	}
+
+	if ( typeof data === 'object' && objectVerificationCallback.call( null, data ) ) {
+		return true;
+	}
+
+	if ( Array.isArray( data ) ) {
+		for ( let i = 0; i < data.length; i++ ) {
+			const type = typeof data[ i ];
+			if ( type !== 'string' && ( type !== 'object' || ! objectVerificationCallback.call( null, data[ i ] ) ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	return false;
```

Comment:
We can use `every` for this check to tidy this up a bit:

```suggestion
	if ( Array.isArray( data ) ) {
		return data.every( ( item ) => {
			typeof item === 'string' || objectVerificationCallback( item )
		 } );
	}

	// Arguably this should fail/throw, because none of our allowed types were encountered.
	return false;
```

Reply:
Updated.